core/src/main/java/eu/fasten/core/dependents/utils/DependencyGraphUtilities.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Set;
+
+/**
+ * Utility functions to construct and (de-)serialize dependency graphs
+ */
+public final class DependencyGraphUtilities {
+
+    private static final Logger logger = LoggerFactory.getLogger(DependencyGraphUtilities.class);
+
+    public static Graph<Revision, DependencyEdge> invertDependencyGraph(Graph<Revision,
+            DependencyEdge> dependencyGraph) {
+        logger.debug("Calculating graph transpose");
+        var startTs = System.currentTimeMillis();
+        var graph = new DefaultDirectedGraph<Revision, DependencyEdge>(DependencyEdge.class);
+        for (var node : dependencyGraph.vertexSet()) {
+            graph.addVertex(node);
+        }
+        for (var edge : dependencyGraph.edgeSet()) {
+            var reversedEdges = new DependencyEdge(edge.target, edge.source, edge.scope, edge.optional, edge.exclusions, edge.type);
+            graph.addEdge(reversedEdges.source, reversedEdges.target, reversedEdges);    // Reverse edges
+        }
+        logger.info("Graph transposed: {} ms", System.currentTimeMillis() - startTs);
+        return graph;
+    }
+
+
+    private static class DefaultPackageVersionSerializer extends Serializer<DefaultArtifactVersion> {
+
+        @Override
+        public void write(Kryo kryo, Output output, DefaultArtifactVersion object) {
+            output.writeString(object.toString());
+        }
+
+        @Override
+        public DefaultArtifactVersion read(Kryo kryo, Input input, Class<? extends DefaultArtifactVersion> type) {
+            var coord = input.readString();
+            try {
+                return type.getConstructor(String.class).newInstance(coord);
+            } catch (Exception e) {
+                logger.warn("Cannot deserialize DefaultArtifactVersion {}", coord);
+                return null;
+            }
+        }
+    }
+
+    private static Kryo setupKryo() throws Exception {
+        var kryo = new Kryo();
+
+        kryo.register(Set.class);
+        kryo.register(HashSet.class);
+        kryo.register(Revision.class);
+        kryo.register(DependencyEdge.class);
+        kryo.register(Dependency.Exclusion.class);
+        kryo.register(Class.forName("eu.fasten.core.dependents.data.Dependency$Exclusion"));
+        kryo.register(java.sql.Timestamp.class);
+        kryo.register(java.util.ArrayList.class);
+        kryo.register(Class.forName("java.util.Collections$UnmodifiableSet"));
+        kryo.register(org.apache.maven.artifact.versioning.DefaultArtifactVersion.class,
+                new DefaultPackageVersionSerializer());
+        kryo.register(org.apache.maven.artifact.versioning.ComparableVersion.class);
+        kryo.register(Class.forName("org.apache.maven.artifact.versioning.ComparableVersion$ListItem"));
+        kryo.register(Class.forName("org.apache.maven.artifact.versioning.ComparableVersion$IntItem"));
+        kryo.register(Class.forName("org.apache.maven.artifact.versioning.ComparableVersion$StringItem"));
```

Comment:
These are for Maven! What should it be for Python & C?

Reply:
According to the [revision call graph format documentation](https://github.com/fasten-project/fasten/wiki/Revision-Call-Graph-format), the Python and C dependency Version Constraints follow the maven format. Therefore, we used the same class employed in Maven Resolver(```org.apache.maven.artifact.versioning.DefaultArtifactVersion```) as the type of Python and C package versions and also  used the method compareTo to make comparsions between them. Thus, on serialization, we need to register the same corresponding Classes, jsut like Maven Resolver.